### DISCUSSION TITLE
How to properly deserialize data

### DISCUSSION THREAD

**Discussion by fairking** [wailsapp/wails#2149] [Q&A] [answered]
Hi guys,

I have a question. I am trying to get a proof of concept working and have some issue with deserializing data.
The following example:
frontend:
```js
Request({ route: 'Store/GetList', body: { page: 1, page_size: 5, search: 'foo' } })
        .then(result => {
            this.responseText = result;
        })
        .catch(error => {
            this.responseText = 'Error: ' + error;
        });
```

app.go:
```go
import { "github.com/mitchellh/mapstructure" }

// Request model
type RequestModel struct {
	Route string
	Body interface{}
}

// Response model
type ResponseModel struct {
	Response interface{}
	Error string
}

type QueryModel struct {
	Page int
	PageSize int
	Search string
}

// Request from the frontend
func (a *App) Request(data *RequestModel) (*ResponseModel, error) {
	switch data.Route {
		case "Store/GetList":
			var query QueryModel
			mapstructure.Decode(query, data.Body)
			fmt.Println(data.Body)
			fmt.Println(query)
			return &ResponseModel{ Response: query }, nil
		default:
			return nil, fmt.Errorf("the route '%s' not found", data.Route)
	}
}
```

Gives me the following result in console (see `fmt.Println(data.Body) fmt.Println(query)`):
```
map[page:1 page_size:5 search:foo] // data.Body
{0 0 } // query
```

Does anyone know why the `query` has an empty result? 
Should `map[string]interface` be properly deserialized?
Am I doing something wrong?

**Comment by fairking** [comment, marked as answer]
Ok, I have found a solution.
1. The json must be in format `pageSize` instead of `page_size`
2. The `mapstructure.Decode(data.Body, &query)` instead of `mapstructure.Decode(query, data.Body)`

**Reply by leaanthony** [reply]
It's a good idea to use JSON tags on your structs 👍

**Reply by fairking** [reply]
I am going to consume public services (api) with more generic approach. Instead of having multiple methods in app.go I would like to have just one for such communication. The frontend will be strong typed (ts) anyway. The pure json provided is just to proof my concept with the communication. I will share my results in near future.

The main goal is upon the call `Request()` to get the required service from DI and then generically call the service and its method based on the `route`.
